src/idpi/grib_decoder.py

```diff
@@ -1,26 +1,134 @@
+import numpy as np
+import xarray as xr
 
-def load_data(outds, fields, datafile, chunk_size=10):
-    chunk_arg = {}
-    if chunk_size:
-        chunk_arg = {"chunks": {"generalVerticalLayer": chunk_size}}
 
-    # Note: dataset assignment is based on typeOfLevel in cfgrib
-    dss = cfgrib.open_datasets(
-        datafile,
-        backend_kwargs={"read_keys": ["typeOfLevel", "gridType"]},
-        encode_cf=("time", "geography", "vertical"),
-        **chunk_arg,
+@contextmanager
+def cosmo_grib_defs():
+    """Enable COSMO GRIB definitions."""
+    prefix = os.environ["CONDA_PREFIX"]
+    root_dir = Path(prefix) / "share"
+    paths = (
+        root_dir / "eccodes-cosmo-resources/definitions",
+        root_dir / "eccodes/definitions",
     )
-    for ds in dss:
-        for field in fields:
-            if field in ds:
-                outds[field] = ds[field]
-                if field == "HHL":
-                    outds[field] = outds[field].interpolate_na(dim="generalVertical")
-
-    if any(field not in outds for field in fields):
-        raise RuntimeError("Not all fields found in datafile", fields)
+    for path in paths:
+        if not path.exists():
+            raise RuntimeError(f"{path} does not exist")
+    defs_path = ":".join(map(str, paths))
+    restore = eccodes.codes_definition_path()
+    eccodes.codes_set_definitions_path(defs_path)
+    try:
+        yield
+    finally:
+        eccodes.codes_set_definitions_path(restore)
+
+
+def load_data(
+    params: list[str], datafiles: list[Path], ref_param: str
+) -> dict[str, xr.DataArray]:
+    """Load data from GRIB files.
+
+    Parameters
+    ----------
+    params : list[str]
+        List of fields to load from the data files.
+    datafiles : list[Path]
+        List of files from which to load the data.
+    ref_param : str
+        Parameter to use as a reference for the coordinates.
+
+    Raises
+    ------
+    ValueError
+        if ref_param is not included in params.
+    RuntimeError
+        if not all fields are found in the given datafiles.
+
+    Returns
+    -------
+    dict[str, xr.DataArray]
+        Mapping of fields by param name
+
+    """
+    fs = earthkit.data.from_source("file", [str(p) for p in datafiles])
+
+    if ref_param not in params:
+        raise ValueError(f"{ref_param} must be in {params}")
+
+    hcoords = None
+    metadata = {}
+    level_types = {}
+    data: dict[str, dict[int, np.ndarray]] = {}
+    for f in fs.sel(param=params):
+        param = f.metadata("param")
+        levels = data.setdefault(param, {})
+        levels[f.metadata("level")] = f.to_numpy(dtype=np.float32)
+
+        if param not in level_types:
+            level_types[param] = f.metadata("typeOfLevel")
+
+        if param not in metadata:
+            metadata[param] = f.metadata(
+                namespace=["ls", "geography", "parameter", "time"]
+            )
+
+        if hcoords is None and param == ref_param:
+            hcoords = {
```

Comment:
I think we can safely remove the horizontal coordinates. The design idea was to put all geographical information in a separate constant object, and attach here only information about staggering wrt the reference grid. 
I would start with nothing and later add it, for the time being we do not have a use case for this.

Reply:
On the contrary, I think we need information about the vertical coordinate. While all variables will be on the same horizontal grid, that is not the case for the vertical. 
https://hackmd.io/559zJIyMQSu_Qi2Sw81WRA
https://github.com/MeteoSwiss-APN/icon_data_processing_incubator/pull/56/files#diff-b8b3cc08fe5792c56c02926372a8e41bacbba127344a55635aedcb14d73f7d8eR47

Reply:
The latitude values are needed in the potential vorticity operator. Until we introduce the grid information object I thought that this was the most obvious place to put this information. In this PR, the vertical dimensions have a different name for different grid locations as they did before.

Reply:
I agree that we should do the grid info in a separate PR. But I was under the impression that the horizontal coordinates are not really used, since we seems to construct them directly from the grib rotated grid information: 
https://github.com/MeteoSwiss-APN/icon_data_processing_incubator/pull/59/files#diff-8bc8502e3870d03e9ad32dd60b42006509625d6e368d08306ed697c331946f46R66

Or where in pot vortic it is needed  ?

Reply:
The geographic latitude is needed for the calculation of the Coriolis terms